compiler/noirc_evaluator/src/ssa/opt/unrolling.rs

```diff
@@ -529,166 +971,402 @@ mod tests {
+            jmp b1()
+          b1():
+            v13 = load v1 -> u32
+            v14 = eq v13, v0
+            constrain v13 == v0
+            return
+        }
+        ";
 
-        builder.switch_to_block(b3);
-        let zero = builder.field_constant(0u128);
-        builder.terminate_with_return(vec![zero]);
+        let (ssa, errors) = try_unroll_loops(ssa);
+        assert_eq!(errors.len(), 0, "Unroll should have no errors");
+        assert_eq!(ssa.main().reachable_blocks().len(), 2, "The loop should be unrolled");
 
-        let ssa = builder.finish();
-        assert_eq!(ssa.main().reachable_blocks().len(), 4);
+        assert_normalized_ssa_equals(ssa, expected);
+    }
 
-        // Expected that we failed to unroll the loop
-        let (_, errors) = try_to_unroll_loops(ssa);
-        assert_eq!(errors.len(), 1, "Expected to fail to unroll loop");
+    /// Test that we can unroll the loop in the ticket if we don't have too many iterations.
+    #[test]
+    fn test_brillig_unroll_6470_small() {
+        // Few enough iterations so that we can perform the unroll.
+        let ssa = brillig_unroll_test_case_6470(3);
+        let (ssa, errors) = try_unroll_loops(ssa);
+        assert_eq!(errors.len(), 0, "Unroll should have no errors");
+        assert_eq!(ssa.main().reachable_blocks().len(), 2, "The loop should be unrolled");
+
+        // The IDs are shifted by one compared to what the ACIR version printed.
+        let expected = "
+        brillig(inline) fn __validate_gt_remainder f0 {
+          b0(v0: [u64; 6]):
+            inc_rc v0
+            inc_rc [u64 0, u64 0, u64 0, u64 0, u64 0, u64 0] of u64
+            v3 = allocate -> &mut [u64; 6]
+            store [u64 0, u64 0, u64 0, u64 0, u64 0, u64 0] of u64 at v3
+            v5 = load v3 -> [u64; 6]
+            v7 = array_get v0, index u32 0 -> u64
+            v9 = add v7, u64 1
+            v10 = array_set v5, index u32 0, value v9
+            store v10 at v3
+            v11 = load v3 -> [u64; 6]
+            v13 = array_get v0, index u32 1 -> u64
+            v14 = add v13, u64 1
+            v15 = array_set v11, index u32 1, value v14
+            store v15 at v3
+            v16 = load v3 -> [u64; 6]
+            v18 = array_get v0, index u32 2 -> u64
+            v19 = add v18, u64 1
+            v20 = array_set v16, index u32 2, value v19
+            store v20 at v3
+            jmp b1()
+          b1():
+            v21 = load v3 -> [u64; 6]
+            dec_rc v0
+            return v21
+        }
+        ";
+        assert_normalized_ssa_equals(ssa, expected);
+    }
+
+    /// Test that with more iterations it's not unrolled.
+    #[test]
+    fn test_brillig_unroll_6470_large() {
+        // More iterations than it can unroll
+        let ssa = brillig_unroll_test_case_6470(6);
+        let stats = loop0_stats(&ssa);
+        assert!(!stats.is_small(), "the loop should be considered large");
+
+        let (ssa, errors) = try_unroll_loops(ssa);
```

Comment:
Do we expect the SSA to be the same? Can we just do a deeper equality check against the SSA output itself?

Reply:
Same for the other failure cases?